services/osparc-gateway-server/src/osparc_gateway_server/backend/osparc.py

```diff
@@ -182,3 +205,132 @@ async def do_check_workers(self, workers: List[Worker]) -> List[bool]:
         )
         self.log.debug("<-- worker status returned: %s", f"{ok=}")
         return ok
+
+    async def on_cluster_heartbeat(self, cluster_name, msg):
+        # pylint: disable=no-else-continue, unused-variable, too-many-branches
+        # pylint: disable=too-many-statements
+
+        # HACK: we override the base class heartbeat in order to
+        # dynamically allow for more or less workers depending on the
+        # available docker nodes!!!
+        cluster = self.db.get_cluster(cluster_name)
+        if cluster is None or cluster.target > JobStatus.RUNNING:
+            return
+
+        cluster.last_heartbeat = timestamp()
+
+        if cluster.status == JobStatus.RUNNING:
+            cluster_update = {}
+        else:
+            cluster_update = {
+                "api_address": msg["api_address"],
+                "scheduler_address": msg["scheduler_address"],
+                "dashboard_address": msg["dashboard_address"],
+            }
+
+        count = msg["count"]
+        active_workers = set(msg["active_workers"])
+        closing_workers = set(msg["closing_workers"])
+        closed_workers = set(msg["closed_workers"])
+
+        self.log.info(
+            "Cluster %s heartbeat [count: %d, n_active: %d, n_closing: %d, n_closed: %d]",
+            cluster_name,
+            count,
+            len(active_workers),
+            len(closing_workers),
+            len(closed_workers),
+        )
+
+        # THIS IS THE HACK!!!
+        # original code in dask_gateway_server.backend.db_base
+        max_workers = cluster.config.get("cluster_max_workers")
+        if self.settings.GATEWAY_SERVER_ONE_WORKER_PER_NODE:
+            # cluster_max_workers = len(await get_cluster_information(self.docker_client))
+            # if max_workers != cluster_max_workers:
+            #     unfrozen_cluster_config = {k: v for k, v in cluster.config.items()}
+            #     unfrozen_cluster_config["cluster_max_workers"] = cluster_max_workers
+            #     cluster_update["config"] = unfrozen_cluster_config
+            max_workers = len(await get_cluster_information(self.docker_client))
```

Comment:
I think this is not that bad of a hack, this will even work if we autoscale the physical cluster. My only concern is;
Can we have ```cluster_max_workers``` dynamic at all or does that have some consequences of how the gateway client works. The ```cluster_max_workers```  can be passed as cluster options to the client and I am not sure if this will break any logic client side. In the worst case this information could be updated every time we submit new work?

Reply:
well let's say that what the bad parts are:
1. I actually had to copy the whole ```on_cluster_heartbeat()``` fct from the base class to overwrite it (so at each update of the library we need to back port this, which for now I agree is maybe not tooo bad)
2. ```cluster_max_workers```: This is a bit more complex and here is my analysis so far:
  - so ```cluster_max_workers``` is defined first in the configuration of the gateway (at start time of the gateway)
  - changing that configuration gateway wide dynamically is not an issue, that is why I first tried using a background task for doing this, by periodically getting from docker the list of nodes, and updating it.
  - THE PROBLEM: that ```cluster_max_workers``` gets copied into the cluster configuration (e.g. when the scheduler starts) when the cluster is created. and that configuration is then saved in an internal sqlite database related to the cluster. That information is set into a FrozenDict
  - I tried: to unfreeze the dict and update it by sending that information directly to the internal database and it partly worked, partly failed since some parts of the cluster configuration disappeared. But I would try to make that work maybe in a later phase again
  - What I do now: I bypass the part where the gateway checks the ```cluster_max_workers``` so that it does not start too many worker that will fail (then failing the whole cluster which would be really bad). On the client side, what is missing now is that a message would tell the client it cannot scale more (but actually also sends a 200 with an additional message)